### ISSUE TITLE
Reading back multidimensional output in C++

### ISSUE THREAD

**Issue by mfenton0** [microsoft/onnxruntime#11718]
Hi,

I am attempting to use ORT to run my PyTorch model in C++ code. I am however having trouble understanding how to parse the output since each output tensor is multidimensional. I find the C++ documentation to be very difficult to parse, and I haven't found a good example similar to my use-case. 

The model has 3 output nodes, with shapes [B, N, N, N], [B], and [B, N, N] where B = batch size and N is a hyperparameter of my network. In principle, N can be fixed, though I see in my output that ONNX has exported it as a dynamic axis - my output shapes are all [-1] even if I do not specify dynamic axes in the export (am I able to enforce this?). In the current case I am using N=20.

My conversion from PyTorch to Onnx looks like this:

```
torch.onnx.export(model,
                 args=(dummy_input[0], dummy_input[1]),
                 f="spanet.onnx",
                 verbose=True,
                 input_names=input_names,
                 output_names=output_names,
                 export_params=True,
                  opset_version=12,
                  dynamic_axes={'features' : {0: 'batch_size'},
                                'mask' : {0: 'batch_size'},
                                'thpred' : {0 : 'batch_size'},
                                'thpresence' : {0 : 'batch_size'},
                                'thmask' : {0 : 'batch_size'},
                                'tlpred' : {0 : 'batch_size'},
                                'tlpresence' : {0 : 'batch_size'},
                                'tlmask' : {0 : 'batch_size'},
                               }
                 )
```

My question is, how do I read back the output? Currently I am trying to do this like so:

  ```
  std::vector<Ort::Value> output_tensors = session.Run(
        input_node_names, input_tensors, output_node_names, Ort::RunOptions{nullptr});

    // now read the output
    for (int j=0; j < 3; ++j){
        auto thpred = output_tensors.at(j).GetTensorMutableData<float>();
        for (int i=0; i < 100; ++i){
            std::cout << thpred[i] << " \t ";
            if (i % 200 == 199) std::cout << std::endl;
        }
    
    std::cout << std::endl << std::endl;
    }
```

where I loop 100 times arbitrarily to investigate what is coming back and use batch size of 2. However, I'm confused about how to interpret what comes out. In what order does the 1-d array that I have called "thpred" unroll the 4-D output of the first tensor?

I appreciate any help anyone is able to provide.

Cheers,
Mike

**Comment by yuslepukhin** [comment]
Ort::Value contains (this case) a Tensor. Tensor contains a shape which can be used to interpret it. However, the buffer contained in a tensor is a contagious piece of memory. The length of the buffer is a product of shape dimensions.

**Comment by mfenton0** [comment]
Right. So, how do I access the tensor stored in the Ort::Value? All the examples I've seen just use float*. Is there some way I can access the tensor using the usual indices?

**Comment by RyanUnderhill** [comment]
If your shape is [B, N, N] then you can pretend that the float* returned is a pointer to a 'float [B][N][N]' array, the data is arranged in memory the same way as the C structure would be.

For dynamic sizes you'll need to do the array index math yourself. If you only want single values, there is a utility API TensorAt: https://onnxruntime.ai/docs/api/c/struct_ort_api.html#a067cabac8cbc83133431f3f0abbc1c80  though this will be very inefficient if you're using it in a loop.

**Comment by yuslepukhin** [comment]
> output_tensors.at(j).GetTensorMutableData<float>();

`output_tensors.at(j).GetTensorMutableData<float>();` this above returns the pointer to a buffer. You might want to call a non-modifying method.

**Comment by mfenton0** [comment]
@yuslepukhin , do you have a specific suggestion for a better way to retrieve the data?

**Comment by RyanUnderhill** [comment]
It's more conceptually correct to call the `.GetTensorData<float>()` method, but they both currently call the same C API method (GetTensorMutableData). This might change in the future if we add a read only version.

**Comment by CarlPoirier** [comment]
If you use the CUDA (Probably TensorRT too) execution provider, and Ort::IoBinding, you get the device pointer and you can transfer the data to the host using cudaMemCpy. Here's an example (untested) for your output tensor of size [B]:

```
Ort::IoBinding binding(*m_ortSession);
binding.BindInput("input", input_tensor);
binding.BindOutput("output", memory_info);

m_ortSession->Run(Ort::RunOptions{ nullptr }, binding);
std::vector<Ort::Value> output_tensors = binding.GetOutputValues();

auto dynamicAxisOutput = output_tensors[0].GetTensorTypeAndShapeInfo().GetShape()[0]; // Retrieve batch size from output
std::vector<float> output(dynamicAxisOutput);
cudaMemcpy(output.data(), output_tensors[0].GetTensorData<float>(), sizeof(float) * dynamicAxisOutput, cudaMemcpyDeviceToHost);
```
